### ISSUE TITLE
How did you handle the multi-level insertion race issue without I-nodes?

### ISSUE THREAD

**Issue by eternaleye** [vorner/contrie#1]
From the paper:

> Intuitively, a concurrent insertion operation could start by locating the internal node it needs to modify and then create a copy of that node with both the bitmap and the array updated with a reference to the key being inserted. A reference to the newly created node could then be written into the array of the parent node using the CAS instruction. Unfortunately, this approach does not work. The fundamental problem here is due to races between an insertion of a key into some node _C1_ and an insertion of another key into its parent node _C2_. One scenario where such a race happens is shown in Figure 1. Assume we have a hash trie from the Figure 1A and that a thread _T1_ decides to insert a key _k5_ into the node _C2_ and creates an updated version of _C2_ called _C2′_. It must then do a CAS on the first entry in the internal node _C3_ with the expected value _C2_ and the new value _C2′_. Assume that another thread _T2_ decides to insert a key _k4_ into the node _C1_ before this CAS. It will create an updated version of _C1_ called _C1′_ and then do a CAS on the first entry of the array of _C2_ – the updated node _C1′_ will not be reachable from the updated node _C2′_. After both threads complete their CAS operations, the trie will correspond to the one shown in Figure 1B, where the dashed arrows represent the state of the branches before the CASes. The key _k4_ inserted by the thread _T2_ is lost.

> We solve this problem by introducing indirection nodes, or I-nodes, which remain present in the Ctrie even as nodes above and below  change. The CAS instruction is performed on the  I-node instead of on the internal node array. We show that this eliminates the race between insertions on different levels.
</blockquote>

I haven't found anything in the implementation that prevents this interleaving; the paper avoids it by dint of the I-nodes forcing such insertions to conflict on their CAS of the I-node, causing one of them to have to retry.

**Comment by vorner** [comment]
I haven't done a formal proof, only some intuitive reasoning, so I might have overlooked something.

However I believe I don't have to handle it because in my implementation it *doesn't happen* in the first place. The problem in the original version arises from the fact one node might need to be replaced by a bigger version when adding a child. If some other thread updates a pointer in the dying node, the update is lost with it.

One of the simplifications I've decided to do is I don't compress the nodes. They start as full-sized already, full of NULL pointers. When adding another child to them, I don't replace the node, I just CaS one of the NULLs with the pointer to the new child. I don't have dying nodes when adding children, so if some other thread updates some other pointer in the same array, nothing is lost (and if two threads try to update the same element, one of them fails on the CaS).

In other words I don't need an I-node to remain in the trie because the node above it never change and the node itself doesn't change. The node itself is therefore able to fulfil the same role implicitly.

Does it make sense, or am I fooling myself?

**Comment by eternaleye** [comment]
Mm, yeah, I see what you're getting at - I suppose I hadn't really internalized that you don't to _any_ CoW up the tree, and that you don't support snapshots at all, rather than the lesser property that "iterators are not snapshots".

On the other hand, that does mean that supporting the full data structure would basically be a full rewrite, rather than an incremental change :/

I'd been hoping it'd be the latter, because the full ctrie interface is what I'd actually find useful, or at-least some way of performing snapshots.

**Comment by vorner** [comment]
> On the other hand, that does mean that supporting the full data structure would basically be a full rewrite, rather than an incremental change :/

Code wise, yes. But now I've done this, it would be much easier for me to do. I needed this stepping stone to actually understand the article and some of the nuances and problems. Nevertheless, I believe even this version might be useful to somebody. I haven't yet decided if I really want to implement the full version eventually, though.

> TBH, I'm not sure I'd call contrie a ctrie implementation, since the snapshot behavior is so central as to be in the title of the original paper.

I think I haven't actually called it a ctrie implementation anywhere (if I did, please point me to the place, I'll fix that). I've only said it was *inspired* by the article, which is true. And the name contrie is „concurrent trie“, which this implementation certainly is (c in ctrie also means concurrent, but…).

**Comment by vorner** [comment]
I believe the original question/issue here has been answered, so I'll close this. If you have anything related to that, don't hesitate to reopen or open a new one.

As for the support for snapshots, I'm going to open a separate issue for that, but I have no idea when or if I'll get around to it.
